File: src/java/3_StructuralPatterns/2_Bridge/BridgeExercise.java
package com.activemesa.structural.bridge.exercise;

//abstract class Shape
//{
//  public abstract String getName();
//}
//
//class Triangle extends Shape
//{
//  @Override
//  public String getName()
//  {
//    return "Triangle";
//  }
//}
//
//class Square extends Shape
//{
//  @Override
//  public String getName()
//  {
//    return "Square";
//  }
//}
//
//class VectorSquare extends Square
//{
//  @Override
//  public String toString()
//  {
//    return String.format("Drawing %s as lines", getName());
//  }
//}
//
//class RasterSquare extends Square
//{
//  @Override
//  public String toString()
//  {
//    return String.format("Drawing %s as pixels", getName());
//  }
//}

// imagine VectorTriangle and RasterTriangle are here too

interface Renderer
{
  public String whatToRenderAs();
}

abstract class Shape
{
  private Renderer renderer;
  public String name;

  public Shape(Renderer renderer)
  {
    this.renderer = renderer;
  }

  @Override
  public String toString()
  {
    return String.format("Drawing %s as %s",
      name, renderer.whatToRenderAs());
  }
}

class Triangle extends Shape
{
  public Triangle(Renderer renderer)
  {
    super(renderer);
    name = "Triangle";
  }
}

class Square extends Shape
{
  public Square(Renderer renderer)
  {
    super(renderer);
    name = "Square";
  }
}

class RasterRenderer implements Renderer
{

  @Override
  public String whatToRenderAs()
  {
    return "pixels";
  }
}

class VectorRenderer implements Renderer
{
  @Override
  public String whatToRenderAs()
  {
    return "lines";
  }
}

File: src/js/3_StructuralPatterns/2_Bridge/bridge_exercise.js
// class Shape
// {
//   constructor(name)
//   {
//     this.name = name;
//   }
// }
//
// class Triangle extends Shape
// {
//   constructor()
//   {
//     super('triangle');
//   }
// }
//
// class Square extends Shape
// {
//   constructor()
//   {
//     super('square');
//   }
// }
//
// class VectorSquare extends Square
// {
//   toString()
//   {
//     return `Drawing square as lines`;
//   }
// }
//
// class RasterSquare extends Square
// {
//   toString()
//   {
//     return `Drawing square as pixels`;
//   }
// }

// imagine VectorTriangle and RasterTriangle are here too

class Shape
{
  constructor(renderer, name=null)
  {
    this.renderer = renderer;
    this.name = name;
  }

  toString()
  {
    return `Drawing ${this.name} as ${this.renderer.whatToRenderAs}`;
  }
}

class Triangle extends Shape
{
  constructor(renderer)
  {
    super(renderer, 'triangle');
  }
}

class Square extends Shape
{
  constructor(renderer)
  {
    super(renderer, 'square');
  }
}

class RasterRenderer
{
  get whatToRenderAs()
  {
    return 'pixels';
  }
}

class VectorRenderer
{
  get whatToRenderAs()
  {
    return 'lines';
  }
}

describe('facade', function()
{
  it('render vector square', function()
  {
    let sq = new Square(new VectorRenderer());
    expect(sq.toString()).toEqual('Drawing square as lines');
  });

  it('render raster triangle', function()
  {
    let sq = new Triangle(new RasterRenderer());
    expect(sq.toString()).toEqual('Drawing triangle as pixels');
  });
});